qutebrowser/misc/sql.py

```diff
@@ -168,82 +171,212 @@ def raise_sqlite_error(msg, error):
+            raise_sqlite_error(msg, error)
+
+        version_int = self.query('pragma user_version').run().value()
+        self._user_version = UserVersion.from_int(version_int)
+
+        if self._user_version.major > self._USER_VERSION.major:
+            raise KnownError(
+                "Database is too new for this qutebrowser version (database version "
+                f"{self._user_version}, but {self._USER_VERSION.major}.x is supported)")
+
+        if self.user_version_changed():
+            # Enable write-ahead-logging and reduce disk write frequency
+            # see https://sqlite.org/pragma.html and issues #2930 and #3507
+            #
+            # We might already have done this (without a migration) in earlier versions,
+            # but as those are idempotent, let's make sure we run them once again.
+            self.query("PRAGMA journal_mode=WAL").run()
+            self.query("PRAGMA synchronous=NORMAL").run()
+
+    def qSqlDatabase(self) -> QSqlDatabase:
+        """Return the wrapped QSqlDatabase instance."""
+        database = QSqlDatabase.database(self._path, open=True)
+        if not database.isValid():
+            raise BugError('Failed to get connection. Did you close() this Database '
+                           'instance?')
+        return database
+
+    def query(self, querystr: str, forward_only: bool = True) -> 'Query':
+        """Return a Query instance linked to this Database."""
+        return Query(self, querystr, forward_only)
+
+    def table(self, name: str, fields: List[str],
+              constraints: Optional[Dict[str, str]] = None,
+              parent: Optional[QObject] = None) -> 'SqlTable':
+        """Return a SqlTable instance linked to this Database."""
+        return SqlTable(self, name, fields, constraints, parent)
+
+    def user_version_changed(self) -> bool:
+        """Whether the version stored in the database differs from the current one."""
+        return self._user_version != self._USER_VERSION
+
+    def upgrade_user_version(self) -> None:
+        """Upgrade the user version to the latest version.
+
+        This method should be called once all required operations to migrate from one
+        version to another have been run.
+        """
+        log.sql.debug(f"Migrating from version {self._user_version} "
+                      f"to {self._USER_VERSION}")
+        self.query(f'PRAGMA user_version = {self._USER_VERSION.to_int()}').run()
+        self._user_version = self._USER_VERSION
+
+    def close(self) -> None:
+        """Close the SQL connection."""
+        database = self.qSqlDatabase()
+        database.close()
+        sip.delete(database)
+        QSqlDatabase.removeDatabase(self._path)
+
+    def transaction(self) -> 'Transaction':
+        """Return a Transaction object linked to this Database."""
+        return Transaction(self)
+
+    @classmethod
+    def version(cls) -> str:
```

Comment:
Should probably be a function in this module rather than a class method? I'm thinking that the sqlite version is more of a generic SQL thing, not really related to the database class (or to a particular database).

Reply:
Do you mean a function that doesn't use the the `Database` class at all? I also tried that. I figured it would be faster since we don't need much of the logic in `Database.__init__` for this use case; but the time difference was practically non existent and it forced me to duplicate the code in `Database .__close__` and the remaining code that we do need from `Database.__init__`. We could still do it if you prefer.

If you mean a function that uses the `Database` class then I think that makes more sense as a class method.

Reply:
I did mean a function using `Database`. I typically only use class methods for alternative constructors (things like `GreasemonkeyScript.parse` or `KeyInfo.from_event`.

IMHO `sql.Database.version()` sounds like it'd give you the version of a database, i.e. it sounds a bit like I'd return `_USER_VERSION` or something - but we're only using the `Database` class as an implementation detail here, and actually get the sql(ite) version.

However, no hard feelings if you prefer to keep it as-is.

Reply:
> IMHO sql.Database.version() sounds like it'd give you the version of a database, i.e. it sounds a bit like I'd return _USER_VERSION or something - but we're only using the Database class as an implementation detail here, and actually get the sql(ite) version.

I'd say that `sql.Database().version()` sounds like that, while `sql.Database.version()` is more like the version of something that is shared between all `Database` instances, just like the library version. I do agree, though, that the use of the class here is just an implementation detail.

I don't have strong feelings for either way, so no worries, I can change it. It's probably something that stuck with me from my Java days: everything's either an instance method or a static method :sweat_smile: